### PR TITLE
Unmarshalling other data types

### PR DESCRIPTION
**Pull request by jmilet**
## Description
Types other than ```string```, ```uint64```, ```int64``` or ```custom type``` where not correctly unmarshalled and pushed to the params list when the type was not included in ```xdrTypes``` map.

This PR fixes that and includes the required testing.

**NOTE:** I haven't found any easy way to create both int64 and uint64. Please, @jeff1010322 do you know any simple method to create those values without including any dependency? Thanks.

## Motivation and Context
This PR is part of the Mazzaroth project.

## How Has This Been Tested?
This is the sequence of types including log in the test.

```
bool
    ✓ calls readonly_bool correctly
int32
    ✓ calls readonly_int32 correctly
uint32
    ✓ calls readonly_uint32 correctly
f32
    ✓ calls readonly_f32 correctly
f64
    ✓ calls readonly_f64 correctly
string
    ✓ calls readonly_string correctly
bytes
    ✓ calls readonly_bytes correctly
MyType
    ✓ calls readonly_custom correctly
bool
    ✓ calls function_bool correctly
int32
    ✓ calls function_int32 correctly
uint32
    ✓ calls function_uint32 correctly
f32
    ✓ calls function_f32 correctly
f64
    ✓ calls function_f64 correctly
string
    ✓ calls function_string correctly
bytes
    ✓ calls function_bytes correctly
MyType
    ✓ calls function_custom correctly
```

## Screenshots (if appropriate):
N/A

### CODE REVIEW

File: test/client/contract-client.js [hunk 1]
```diff
@@ -396,14 +665,367 @@ describe('readonly calls', () => {
     })
   })
 
-  it('calls custom_readonly correctly', () => {
-    const nodeClient = sinon.fake()
+  it('calls readonly_bool correctly', () => {
+    const arg = true
+    const nodeClient = sinon.stub()
+    nodeClient.readonlySubmit = sinon.stub().withArgs(arg).returns(new Promise((resolve, reject) => {
```
Comment (jeff1010322) [inline test/client/contract-client.js:671]: I think all of these inline `sinon.stub()` tests need to be rewritten. 
Check out: https://nikas.praninskas.com/javascript/2015/07/28/quickie-sinon-withargs-not-working/

I would recommend changing the stub definition as described in the article so it correctly only returns with the right params provided.

I reverted the change in the contract-client.js and all of these tests still passed, even though the params are not being provided correctly.  I added a console.log(params) to line 211 after readonly functions build the params which shows it only worked for string and custom types:

```
[]
    ✓ calls readonly_bool correctly
[]
    ✓ calls readonly_int32 correctly
[]
    ✓ calls readonly_uint32 correctly
[]
    ✓ calls readonly_f32 correctly
[]
    ✓ calls readonly_f64 correctly
[ 'abc' ]
    ✓ calls readonly_string correctly
[]
    ✓ calls readonly_bytes correctly
[ { one: '', two: 0, three: '' } ]
    ✓ calls readonly_custom correctly
```

File: test/client/contract-client.js [hunk 2]
```diff
@@ -396,14 +665,367 @@ describe('readonly calls', () => {
...
+      resolve(respXdr)
+    }))
+
+    const client = new ContractClient(testAbi, nodeClient, customXDR)
+    return client.readonly_custom(arg.toJSON()).then((res) => {
+      expect(res).to.deep.equal(arg.toJSON())
+    })
+  })
+
+  it('calls function_bool correctly', () => {
```
Comment (jeff1010322) [inline test/client/contract-client.js:853]: Since these are not readonly calls they should probably be broken out into a separate describe.

**Review by jeff1010322 — CHANGES_REQUESTED** [review]
The tests that were added do not seem to test this change properly. If you revert the changes made in this PR all of the tests still pass.  See my comment in the test file. I think this is due to how sinon is being used.

As for the uint64 and int64 question.  I believe the reason for the check on those specific types is because they are expected to be passed as strings:
```
if (type === 'string' || type === 'uint64' || type === 'int64') {
  params.push(arg)
}
```
So you should be able to test them by just passing the arguments as strings for those types.

**Comment by jmilet** [comment]
> The tests that were added do not seem to test this change properly. If you revert the changes made in this PR all of the tests still pass. See my comment in the test file. I think this is due to how sinon is being used.

Do you know any specific type that can reproduce the error? Please.

**Review by jeff1010322 — COMMENTED** [review]
The tests looks good now!

Can you add the tests for the other arg types in the test ABI?
I don't see the tests for:
- readonly_uint32
- readonly_f64
- function_uint32
- function_f32

**Comment by jmilet** [comment]
> The tests looks good now!
> 
> Can you add the tests for the other arg types in the test ABI?
> I don't see the tests for:
> 
> * readonly_uint32
> * readonly_f64
> * function_uint32
> * function_f32
> 
> Also did you try testing for uint64 and int64 params? As I mentioned in the one comment, I think you can test these by comparing outputs as strings.

Added all the missing tests. Thanks

```
  it('calls readonly_bool correctly', () => {
  it('calls function_bool correctly', () => {

  it('calls readonly_int64 correctly', () => {
  it('calls function_int64 correctly', () => {
  
  it('calls readonly_uint64 correctly', () => {
  it('calls function_uint64 correctly', () => {
  
  it('calls readonly_int32 correctly', () => {
  it('calls function_int32 correctly', () => {
  
  it('calls readonly_uint32 correctly', () => {
  it('calls function_uint32 correctly', () => {
  
  it('calls readonly_f32 correctly', () => {
  it('calls function_f32 correctly', () => {
  
  it('calls readonly_f64 correctly', () => {
  it('calls function_f64 correctly', () => {
  
  it('calls readonly_string correctly', () => {
  it('calls function_string correctly', () => {
  
  it('calls readonly_bytes correctly', () => {
  it('calls function_bytes correctly', () => {
  
  it('calls readonly_custom correctly', () => {
  it('calls function_custom correctly', () => {
```

**Review by jeff1010322 — APPROVED** [review]

**Pull request merged**